Add ruff format to format code

ericpre opened the pull request:
Follow up of and based on https://github.com/hyperspy/hyperspy/pull/3299.

Close #2316.

### Progress of the PR
- [x] Run `ruff format` on code,
- [x] add `ruff format` to `pre-commit` config,
- [n/a] update docstring (if appropriate),
- [n/a] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [n/a] add tests,
- [x] ready for review.

ericpre committed:
Run ruff format on tests

ericpre committed:
ruff format

ericpre committed:
Run ruff format on doc

ericpre committed:
Enable `ruff format` in pre-commit hook

ericpre committed:
Add changelog entry

CSSFrancis reviewed (approved):
@ericpre this looks good to me (although it is hard to go the changes in a formatting PR like this). I personally love ruff and so I'm a big fan of adding this to hyperspy.  

One thing that might be nice is to follow the guide [here](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html) which goes through how to not screw up the git blame the first time using black (or ruff)

ericpre committed:
Add ruff format commits to `.git-blame-ignore-revs`

ericpre commented:
> @ericpre this looks good to me (although it is hard to go the changes in a formatting PR like this). I personally love ruff and so I'm a big fan of adding this to hyperspy.

Yes, I split the linting in `ruff check` (#3299) which need manual edits from the `ruff format` which fix the code automatically (this PR). The linting correction can introduce bugs and reviewing but the code formatting shouldn't - in any I committed formatting of the test suite from the rest in case it helps figuring bugs out in the future.
Do you want to review #3299 separately?

> One thing that might be nice is to follow the guide [here](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html) which goes through how to not screw up the git blame the first time using black (or ruff)

Yes, I didn't know about it, this is useful indeed and it works:
https://github.com/ericpre/hyperspy/blame/ruff_format/hyperspy/conftest.py#L18

I added the formatting commits only, not the linting.

CSSFrancis commented:
> Yes, I split the linting in `ruff check` (#3299) which need manual edits from the `ruff format` which fix the code automatically (this PR). The linting correction can introduce bugs and reviewing but the code formatting shouldn't - in any I committed formatting of the test suite from the rest in case it helps figuring bugs out in the future. Do you want to review #3299 separately?

Oh perfect, I've reviewed that as well. That is much easier to follow :)


> Yes, I didn't know about it, this is useful indeed and it works: https://github.com/ericpre/hyperspy/blame/ruff_format/hyperspy/conftest.py#L18

Cool! That seems easy and useful. I've been meaning to go back and do that in pyxem etc.  In this case I figured getting ahead of it is the best plan :)
